exchanges/request/limit_test.go

```diff
@@ -0,0 +1,256 @@
+package request
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"testing/synctest"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"golang.org/x/time/rate"
+)
+
+func TestRateLimit(t *testing.T) {
+	t.Parallel()
+
+	synctest.Test(t, func(t *testing.T) { //nolint:thelper // false positive
+		err := (*RateLimiterWithWeight)(nil).RateLimit(t.Context())
+		assert.ErrorContains(t, err, "nil pointer: *request.RateLimiterWithWeight")
+
+		r := &RateLimiterWithWeight{limiter: rate.NewLimiter(rate.Limit(1), 1)}
+		err = r.RateLimit(t.Context())
+		assert.ErrorIs(t, err, errInvalidWeight, "should return errInvalidWeightCount for zero weight")
+
+		r = NewRateLimitWithWeight(time.Second, 10, 1)
+		start := time.Now()
+		err = r.RateLimit(t.Context())
+		synctest.Wait()
+		elapsed := time.Since(start)
+		require.NoError(t, err, "rate limit must not error")
+		assert.Less(t, elapsed, time.Millisecond*50, "should complete quickly for first request")
```

Comment:
🦍 Optional pedantic nitpick:
This gorilla is a 🍆 , but he thinks we can save some lines in a test file 🙄 
```suggestion
		assert.Less(t, time.Since(start), time.Millisecond*50, "should complete quickly for first request")
		require.NoError(t, err, "rate limit must not error")
```